README.md
# php_srrJSEncrypt

Simple way (unique PHP file) to store your encrypted data in a single text file, not using database at all and encrypting from the client side using javascript AES algorithms. It also generate 6-digit TOTp codes (see below).

Current version: 2020.06.15 - [Changelog](/CHANGELOG.TXT)

## Screenshots

![first access window](/screenshot3.png?raw=true "First access window")

![client side encrypted data](/screenshot.png?raw=true "Client side encrypted data")

![decrypted data](/screenshot2.png?raw=true "Decrypted data")


## Start the project

Copy this project in any PHP 5+ server and call it from browser. Default encryption key "1234". Each time you decrypt, you can change key and re-encrypt and save.
You must remember that the encryption key is not saved ANYWHERE, so you must take care by yourself for not forget it!

## Protection with a md5 password 

To access to the index.php menu file of the interface you will submit a password in the login.php page. So, you must previously embbed in that file the md5 hash of your preferred password, inthe first lines of that PHP file.

## Data files

On the directory `/data` you can put as many data files you need. Then the `index.php` show at home a list of the files included in that directory. The name of those files cannot contain dots or inclined slashes (. / \ ) due to security reasons. It will be recommendable only to use letters, numeric digits and horizontal slashes.

## 2FA TOTp codes generator

if you use a data file to store a JavaScript Object (like the backup export file of the browser extension  [Authenticator](https://github.com/Authenticator-Extension/Authenticator) ) then when you decrypt it appears a list of 6-digit TOTp codes generated by the list of accounts contained in the data.

![2FA TOTp generator](/screenshot4.png?raw=true "TOTp 6-digit codes")

This picture below show an example of the kind of data. The app is ready to understand the exported backup of that Firefox addon, but you can write mannually the list of accounts whenever you include for each one: a label "account" and a "secret" key, at least. The other properties are optional and useless for this app.

![2FA needed data](/screenshot5.png?raw=true "2FA needed data")

The app doesn't automatically regenerate the codes neither show a countdown indicator. You must click over `regenerate` button to update 6-digit codes.

## Use, improve and share

You're welcome to use it. I use it since long time ago for save all my important private data, hosting a copy of this project on one of my servers.
Please, if you make any improvement, request a pull ;)

## Credit

Created by [caos30](https://github.com/caos30)


CHANGELOG.TXT
================================
   php_srrJSEncrypt CHANGELOG
================================

== v 1.3 [2020-06-15]
- Added keyboard shortcuts on index.php, to use TAB key to move on the button list
- Added keyboard shortcuts on the TOTP codes generator: key R to refresj the codes on screen for each code button, and TAB key to move on the button list.

== v 1.2 [2019-03-09]
- Added store for 2FA TOTP codes and generator of 6-digit temporal codes.
- Added a login page asking a password (embbeded on login.php as md5 hash)




login.php
<?php

	if (session_status() == PHP_SESSION_NONE) session_start();

	$version = '2020.06.15';
	
// == validate password
	if (!empty($_POST['pwd']) && md5($_POST['pwd'])=="1d91331df1d29c73ddf48f3123418f46"){
		$_SESSION['LOGGED'] = '1';
		echo "<script>document.location='index.php';</script>";
		die();
	}

?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns='http://www.w3.org/1999/xhtml'><head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<title> JS-ENCRYPT </title>
<meta name="title" content="JS-ENCRYPT" />
<meta name="description" content="" />
<meta name="keywords" content="" />
<meta http-equiv="pragma" content="no-cache" />
<meta name="revisit-after" content="7 days" />
<meta name="robots" content="nofollow" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<!-- favicons generated at https://realfavicongenerator.net -->

    <link rel="apple-touch-icon" sizes="180x180" href="images/favicon/apple-touch-icon.png">
    <link rel="icon" type="image/png" href="images/favicon/favicon-32x32.png" sizes="32x32">
    <link rel="icon" type="image/png" href="images/favicon/favicon-16x16.png" sizes="16x16">
    <link rel="manifest" href="images/favicon/manifest.json">
    <link rel="mask-icon" href="images/favicon/safari-pinned-tab.svg" color="#5bbad5">
    <link rel="shortcut icon" href="images/favicon/favicon.ico">
    <meta name="apple-mobile-web-app-title" content="JSEncrypt">
    <meta name="application-name" content="JSEncrypt">
    <meta name="msapplication-config" content="images/favicon/browserconfig.xml">
    <meta name="theme-color" content="#ffffff">

	<link href="images/estilos_2020-01-25.css" rel="stylesheet" type="text/css" />

	<style>
		.file_list{text-align:center;margin:2rem;}
		.file_list a{display:block;padding:1rem 2rem;margin:2rem auto;max-width:400px; text-decoration:none;background-color:#000;color:#eee;font-size:1.5rem;}
		.file_list a:hover{box-shadow:0px 0px 7px 1px #fff;}
	</style>

</head>
<body>

	<?php if (!empty($_POST['pwd'])){ ?>
		<p style="color:yellow;text-align:center;">Contraseña errónea.</p>
	<?php } ?>

	<form method="post" id="login_form" style="margin-top:5rem;margin-bottom:7rem;">
		<p style="text-align:center;"><b>Contraseña</b><br /></p>
		<p style="text-align:center;">
			<input type="password" name="pwd" value=""
				style="line-height:3rem;font-size:1.5rem;padding:0 1rem;background-color:#333;color:yellow;border:1px #555 solid;"
				onkeypress="if(getKeyCode(event)==13){$('#login_form').submit();}"
            autocomplete="off" />
		</p>
	</form>

	<p id="signature"><a href="https://github.com/caos30/php_srrJSEncrypt" target="_blank">php_srrJSEncrypt - <?= $version ?></a> (C) 2015-2020 GPL v2 license</p>

</body>
</html>


index.php
<?php /*
 *
 *  php_srrJSEncrypt :: an easy not-database depending system for store encrypted data using only javascript on client side
    Copyright (C) 2015-2019  caos30

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License along
    with this program; if not, write to the Free Software Foundation, Inc.,
    51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */ ?>
<?php

if (session_status() == PHP_SESSION_NONE) session_start();
if (empty($_SESSION['LOGGED'])) echo "<script>document.location='login.php';</script>";

$version = '2020.06.15';

// = get list of data files
	$scan = scandir(dirname(__FILE__).'/data/');
	$files = array();
	if (is_array($scan) && count($scan)>0){
		foreach ($scan as $name){
			if ($name!='.' && $name!='..') $files[] = $name;
		}
	}


?>

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN" "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
<html xmlns='http://www.w3.org/1999/xhtml'><head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<title> JS-ENCRYPT </title>
<meta name="title" content="JS-ENCRYPT" />
<meta name="description" content="" />
<meta name="keywords" content="" />
<meta http-equiv="pragma" content="no-cache" />
<meta name="revisit-after" content="7 days" />
<meta name="robots" content="nofollow" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<!-- favicons generated at https://realfavicongenerator.net -->

    <link rel="apple-touch-icon" sizes="180x180" href="images/favicon/apple-touch-icon.png">
    <link rel="icon" type="image/png" href="images/favicon/favicon-32x32.png" sizes="32x32">
    <link rel="icon" type="image/png" href="images/favicon/favicon-16x16.png" sizes="16x16">
    <link rel="manifest" href="images/favicon/manifest.json">
    <link rel="mask-icon" href="images/favicon/safari-pinned-tab.svg" color="#5bbad5">
    <link rel="shortcut icon" href="images/favicon/favicon.ico">
    <meta name="apple-mobile-web-app-title" content="JSEncrypt">
    <meta name="application-name" content="JSEncrypt">
    <meta name="msapplication-config" content="images/favicon/browserconfig.xml">
    <meta name="theme-color" content="#ffffff">

	<link href="images/estilos_2020-06-15.css" rel="stylesheet" type="text/css" />

</head>
<body>

	<div class="file_list">
	<?php 	$ii=0; 
			foreach ($files as $file){ 
				if ($file=='.htaccess') continue;
				$ii++; ?>
				<a href="open.php?file=<?= urlencode($file) ?>" id="shrtct_<?= $ii ?>" target="_blank"><span><?= $ii ?></span><?= str_replace(array('-','_'),' ',$file) ?></a>
	<?php 	} ?>
	</div>

	<p id="signature"><a href="https://github.com/caos30/php_srrJSEncrypt" target="_blank">php_srrJSEncrypt - <?= $version ?></a> (C) 2015-2020 GPL v2 license</p>
	
	<script>
		// == detect if the a numeric (1-9) keyboard is pressed (normal or numeric keyboard)
		// == to trigger the opening of the corresponding file
		var shrtct_ii = 0;
		document.onkeyup = function(e) {
			var key = e.which || e.keyCode;
			
			if (key==9){
				shrtct_ii++;
				if (shrtct_ii==1){
					document.getElementById('shrtct_'+shrtct_ii).focus();
				}else if (shrtct_ii > <?= $ii ?>){
					shrtct_ii = 1;
					document.getElementById('shrtct_'+shrtct_ii).focus();
				}else{
					// it's not necessary to change focus because the web browser do it by us
				}
			}
			
			//alert(key);
			return;
			/*
			var number = '';
			if (key == 49 || key == 97  ) { number = '1'; }else 
			if (key == 50 || key == 98  ) { number = '2'; }else 
			if (key == 51 || key == 99  ) { number = '3'; }else 
			if (key == 52 || key == 100 ) { number = '4'; }else 
			if (key == 53 || key == 101 ) { number = '5'; }else 
			if (key == 54 || key == 102 ) { number = '6'; }else 
			if (key == 55 || key == 103 ) { number = '7'; }else 
			if (key == 56 || key == 104 ) { number = '8'; }else 
			if (key == 57 || key == 105 ) { number = '9'; }
			if (number!='') document.getElementById('shrtct_'+number).focus();
			/*
				keyboard special keys: e.ctrlKey , e.altKey , e.shiftKey
				example: if (e.ctrlKey && key == 66) {}
			*/
		};
	</script>

</body>
</html>
